packages/back-end/src/events/handlers/webhooks/EventWebHookNotifier.ts

```diff
@@ -0,0 +1,279 @@
+  ) {
+    if (jobDefined) return;
+
+    this.agenda.define<EventWebHookJobData>(
+      "eventWebHook",
+      EventWebHookNotifier.handleAgendaJob
+    );
+    jobDefined = true;
+  }
+
+  /**
+   * Enqueue the job to be performed immediately asynchronously in Agenda
+   */
+  async enqueue(): Promise<void> {
+    const job = this.agenda.create<EventWebHookJobData>("eventWebHook", {
+      ...this.options,
+      retryCount: 0,
+    });
+    job.unique({
+      "data.eventId": this.options.eventId,
+      "data.eventWebHookId": this.options.eventWebHookId,
+    });
+    job.schedule(new Date());
+    await job.save();
+  }
+
+  /**
+   * This is the entry point for when the job executes
+   * @param job
+   * @private
+   */
+  private static async handleAgendaJob(
+    job: Job<EventWebHookJobData>
+  ): Promise<void> {
+    const { eventId, eventWebHookId } = job.attrs.data;
+
+    const event = await getEvent(eventId);
+    if (!event) {
+      // We should never get here.
+      throw new Error(
+        `EventWebHookNotifier -> ImplementationError: No event for provided ID ${eventId}`
+      );
+    }
+
+    const eventWebHook = await getEventWebHookById(eventWebHookId);
+    if (!eventWebHook) {
+      // We should never get here.
+      throw new Error(
+        `EventWebHookNotifier -> ImplementationError: No webhook for provided ID: ${eventWebHookId}`
+      );
+    }
+
+    const organization = await findOrganizationById(event.organizationId);
+    if (!organization) {
+      throw new Error(
+        `EventWebHookNotifier -> ImplementationError: No organization for ID: ${event.organizationId}`
+      );
+    }
+
+    const savedGroupMap = await getSavedGroupMap(organization);
+    const payload = getPayloadForNotificationEvent({
+      event: event.data,
+      organization,
+      savedGroupMap,
+    });
+
+    const webHookResult = await EventWebHookNotifier.sendDataToWebHook({
+      payload,
+      eventWebHook,
+    });
+
+    switch (webHookResult.result) {
+      case "success":
+        return EventWebHookNotifier.handleWebHookSuccess(
+          job,
+          webHookResult,
+          organization.id,
+          payload
+        );
+
+      case "error":
+        return EventWebHookNotifier.handleWebHookError(
+          job,
+          webHookResult,
+          organization.id,
+          payload
+        );
+    }
+  }
+
+  /**
+   * This function makes the post request to the given event web hook with the provided payload,
+   * signing it.
+   * @param payload
+   * @param eventWebHook
+   */
+  private static async sendDataToWebHook<DataType>({
+    payload,
+    eventWebHook,
+  }: {
+    payload: DataType;
+    eventWebHook: EventWebHookInterface;
+  }): Promise<EventWebHookResult> {
+    try {
+      const { url, signingKey } = eventWebHook;
+
+      const signature = getEventWebHookSignatureForPayload({
+        signingKey,
+        payload,
+      });
+
+      const res = await fetch(url, {
```

Comment:
I don't know if fetch is flexible enough for webhooks.  What if someone points to a URL that has a 10GB response?  What if it's a long HTTP connection that stays open for 5 minutes?  Those are basically DOS attacks that could crash the GrowthBook containers.

Ideally we can have a reasonable timeout and limit on the response length.  Once we read the first X kilobytes of response data, we can abort the request.

The same issue applies to the existing webhooks, so I'm fine leaving this as-is and addressing in a follow-on PR.

Reply:
With regards to long-running responses, time-wise, I've implemented a solution with AbortController to abort requests that run longer than 10 seconds. We can configure this as needed. We shouldn't need 10 seconds if consumers are implementing web hooks correctly (i.e. responding immediately, then handling requests), but it's at 10 to be flexible. Let me know if you want me to change the value: https://github.com/growthbook/growthbook/pull/760/commits/9ce063ab9395f1cb880f642b7242cae08b919640 

The example was also updated to include an easy way to simulate network latency using a delay: https://github.com/growthbook/examples/pull/18/commits/586e7564a921aa205338b0a5ffe88899c2f3344e

With regards to content size, I will address that in a follow up commit.

Reply:
I have updated the implementation to include checks for content size, truncating the response and aborting if it's larger than the provided max content size config. You can see the implementation here: https://github.com/growthbook/growthbook/pull/760/files#diff-71a751ed8831bae36cb6703814a3725525bc598f3e2b4370c2f353b1fdc9b6bfR174-R240